### ISSUE TITLE
Make DC/OS CLI Multi Cluster aware

### ISSUE THREAD

**Issue by lloesche** [dcos/dcos-cli#595]
Often times I have more than one DC/OS cluster. I might have separate staging and production systems. It would be nice if instead of having to configure several dcos cli's I could have a single one with multiple DC/OS backends configured and then switch between them like e.g.
`dcos marathon app list --cluster=prod`
`dcos marathon app list --cluster=stage`

Together with another function that I'm going to create a separate issue (https://github.com/dcos/dcos-cli/issues/596) for I could then do things like `dcos marathon backup --cluster=prod | dcos marathon restore --cluster=stage`

**Comment by tamarrow** [comment]
Yes, this is a commonly requested feature. As a workaround for now, you can have multiple config files, each for a different cluster, and use the `DCOS_CONFIG` env variable to override what cluster you want to use. Something like `DCOS_CONFIG=~/.dcos/dcos-dev.toml marathon app list`

**Comment by arun-gupta** [comment]
This would be very useful as multiple clusters would be a common scenario. The current way to configure `dcos` CLI is a three step process:

- `dcos config set core.dcos_url <new-dcos-url>`
- OAuth token
- `dcos auth login`

It needs to be simplified.

**Comment by prein** [comment]
I'm using the workaround as suggested by @tamarrow 
However I'm hitting strange issue when running dcos cli, in docker, from gitlab ci runner (I assume it cleans up after each run).
There are 3 toml configs. I set DCOS_CONFIG to the one I want to use. Then I print 'dcos config show'.  It shows the correct url, but then if as very next command I run any dcos cli command requiring auth, it fails for the reason that it's using another cluster url, from other toml file. I'm so confused to see that happening. 

I run it with strace and it confirmed that it's opening only the correct toml file. No other toml file is touched. And no, I don't have default dcos.toml. Then the wrong url comes out of the blue. So confusing. I must be doing some simple mistake perhaps.

**Comment by tamarrow** [comment]
@prein do you have `DCOS_URL` set? It overrides the url being used.

**Comment by prein** [comment]
Oh, I do have DCOS_URL set indeed, to the wrong one! Dunno how come, yet. Thank you for great hint @tamarrow

**Comment by msabramo** [comment]
One simple thing to add would be to check for a `.dcos` directory in the current directory and prefer that over `~/.dcos`. Then you could have directories for different clusters. OTOH, folks may not always want to map a directory to a cluster. 

Another way to handle it is to add `dcos cluster list` and `dcos cluster switch` commands. Switch could possibly work by just making `~/.dcos/dcos.toml` a symlink that can be switched to point to different files for each cluster.

**Comment by msabramo** [comment]
Or maybe the the directory `~/.dcos` should be a symlink. That way each environment can have different subcommands too.

**Comment by luisdavim** [comment]
I would like to see a similar implementation to this:
http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#cli-multiple-profiles

**Comment by tamarrow** [comment]
This is happening for the next release. I'll update this issue with details once it's merged to master.

**Comment by tamarrow** [comment]
This is now available on master. To setup a cluster run `dcos cluster setup`, to see a list of configured clusters run `dcos cluster list`, to switch to a different cluster run `dcos cluster attach <cluster-name>`. For full command details run `dcos cluster -h` and feel free to ping me for help/problems/etc.

**Comment by bamarni** [comment]
@leviramsey : Indeed it is not safe to setup / attach to different clusters in parallel, the advantage of the current approach though is that one doesn't have to export an environment variable or run a script for each new shell session. At the moment, if you have multiple clusters and multiple DC/OS CLI users (whether it's a user or automated job) they should all use a dedicated `DCOS_DIR` or system user.

To simplify these use cases, what about : 
- supporting an environment variable which would override the attached cluster (`DCOS_CLUSTER=<cluster_name_or_id> dcos [...]`)
- adding a `--no-attach` flag to `dcos cluster setup`

**Comment by leviramsey** [comment]
Yeah, I think supporting an environment variable and having that variable override the attach is probably the best course for preserving predictability and flexibility.

**Comment by bamarni** [comment]
@leviramsey : see the related PR (https://github.com/dcos/dcos-cli/pull/1098) which provides an environment variable that can be used to safely attach to different clusters concurrently.
